sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -144,7 +154,21 @@
             if (options.getExecutorService() != null) {
                 publicClientApplicationBuilder.executorService(options.getExecutorService());
             }
+            if (options.isSharedTokenCacheEnabled()) {
+                try {
+                    publicClientApplicationBuilder.setTokenCacheAccessAspect(
+                            new PersistenceTokenCacheAccessAspect(options.getPersistenceSettings()));
+                } catch (Throwable t) {
+                    String message = "Shared token cache is unavailable in this environment.";
+                    if (sharedTokenCacheCredential) {
+                        throw logger.logExceptionAsError(new CredentialUnavailableException(message, t));
+                    } else {
+                        throw logger.logExceptionAsError(new ClientAuthenticationException(message, null, t));
+                    }
+                }
+            }
             this.publicClientApplication = publicClientApplicationBuilder.build();
```

Comment:
This is not thread safe. If two clients using the same credential call get token at the same time it's possible that multiple threads attempt to initialize `publicClientApplication` at the same time. In this case they might each get different instances and any tokens cached on any instances other than the final one will be lost.

Reply:
Opened an issue #10675

Reply:
Resolving as separate issues opened